vcd/datasource_vcd_resource_list.go

```diff
@@ -0,0 +1,720 @@
+				Optional:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+			},
+			// Parent will be needed for:
+			// * VM (parent: vApp)
+			// * catalogItem (catalog)
+			// * mediaItem (catalog)
+			// * all edge gateway objects (NAT, firewall, lb)
+			// When the parent is org or vdc, they are taken from the regular fields above
+			"parent": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The name of the parent to the resources being retrieved",
+			},
+			"name": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Unique name of the Info",
+			},
+			"resource_type": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Which resource we should list",
+			},
+			"list": &schema.Schema{
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Holds the list of requested resources",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"list_mode": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "name",
+				Description: "How the list should be built",
+				ValidateFunc: validation.StringInSlice([]string{
+					"name",      // The list will contain only the entity name
+					"id",        // The list will contain only the entity ID
+					"href",      // The list will contain only the entity HREF
+					"import",    // The list will contain the terraform import command
+					"name_id",   // The list will contain name + ID for each item
+					"hierarchy", // The list will contain parent names + resource name for each item
+				}, true),
+			},
+			"name_id_separator": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "  ",
+				Description: "Separator for name_id combination",
+			},
+		},
+	}
+}
+
+func orgList(d *schema.ResourceData, meta interface{}) (list []string, err error) {
+	client := meta.(*VCDClient)
+
+	listMode := d.Get("list_mode").(string)
+	nameIdSeparator := d.Get("name_id_separator").(string)
+	orgList, err := client.VCDClient.GetOrgList()
+	if err != nil {
+		return list, err
+	}
+	for _, org := range orgList.Org {
+
+		adminOrg, err := client.GetAdminOrgByName(org.Name)
+		if err != nil {
+			return []string{}, err
+		}
+		switch listMode {
+		case "name", "hierarchy":
+			list = append(list, org.Name)
+		case "id":
+			list = append(list, adminOrg.AdminOrg.ID)
+		case "name_id":
+			list = append(list, org.Name+nameIdSeparator+adminOrg.AdminOrg.ID)
+		case "href":
+			list = append(list, org.HREF)
+		case "import":
+			list = append(list, fmt.Sprintf("terraform import vcd_org.%s %s", org.Name, org.Name))
+		}
+	}
+	return list, err
+}
+
+func externalNetworkList(d *schema.ResourceData, meta interface{}) (list []string, err error) {
+	client := meta.(*VCDClient)
+
+	if !client.VCDClient.Client.IsSysAdmin {
+		return []string{}, nil
```

Comment:
Here I would return better error then expect user to know that he iusing none sys user

Reply:
changed